In [1]:
from inference_and_generation import *


/cephfs/users/bashir/RCoT-Diffusion-LLM/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Using cuda:0


In [2]:
%%time

model_name = "Qwen/Qwen2.5-7B"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
temperature = 0.0
top_p = 0.95
n_samples = 2

metrics = evaluate_auto_regressive_model(
    model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, temperature, top_p, n_samples
    )

Loading checkpoint shards: 100%|██████████| 4/4 [00:00<00:00, 104.61it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5):   0%|          | 0/3 [00:00<?, ?it/s]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Evaluating (total samples: 5):  33%|███▎      | 1/3 [00:14<00:28, 14.46s/it]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Evaluating (total samples: 5):  67%|██████▋   | 2/3 [00:30<00:15, 15.58s/it]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The following generation flags are not valid and may be ignored: 

Saved generations to results/Qwen_Qwen2.5-7B-base_256_0_2_0.0_generations_ar.json
CPU times: user 35.1 s, sys: 11.5 s, total: 46.6 s
Wall time: 55.6 s


In [3]:
%%time

diffusion_model_name = "GSAI-ML/LLaDA-8B-Base"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 32
temperature = 0.0
cfg_scale = 0.0
steps = 32
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2

metrics = evaluate_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples
    )

Loading checkpoint shards: 100%|██████████| 6/6 [00:00<00:00, 176.15it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [00:18<00:00,  6.10s/it]


Saved generations to results/GSAI-ML_LLaDA-8B-Base_256_32_0_2_0.0_generations_testing.json
CPU times: user 22.3 s, sys: 13.3 s, total: 35.6 s
Wall time: 39.4 s


In [4]:
%%time

diffusion_model_name = "GSAI-ML/LLaDA-8B-Base"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 64
temperature = 0.0
cfg_scale = 0.0
steps = 64
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2

metrics = evaluate_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples
    )

Loading checkpoint shards: 100%|██████████| 6/6 [00:00<00:00, 174.36it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [00:37<00:00, 12.58s/it]


Saved generations to results/GSAI-ML_LLaDA-8B-Base_256_64_0_2_0.0_generations_testing.json
CPU times: user 39.2 s, sys: 15.4 s, total: 54.5 s
Wall time: 58 s


In [2]:
%%time

diffusion_model_name = "GSAI-ML/LLaDA-8B-Base"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 256
temperature = 0.0
cfg_scale = 0.0
steps = 256
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2
device = "cuda"
output_dir = "results"

metrics = evaluate_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples, device, output_dir
    )

Loading checkpoint shards: 100%|██████████| 6/6 [00:00<00:00, 174.21it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [02:27<00:00, 49.05s/it]


Saved generations to results/GSAI-ML_LLaDA-8B-Base_256_256_0_2_0.0_generations_testing.json
CPU times: user 2min 17s, sys: 25.8 s, total: 2min 43s
Wall time: 2min 46s


In [2]:
%%time

diffusion_model_name = "GSAI-ML/LLaDA-8B-Base"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 256
temperature = 0.0
cfg_scale = 0.0
steps = 256
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2
use_cache=True
dual_cache=True
threshold = 0.3
factor = 1.0

metrics = evaluate_fast_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples, use_cache, dual_cache, threshold, factor
    )

Loading checkpoint shards: 100%|██████████| 6/6 [00:00<00:00, 169.58it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [00:23<00:00,  7.91s/it]


Saved generations to results/GSAI-ML_LLaDA-8B-Base_256_256_0_2_0.0_generations_testing_fast_dllm.json
CPU times: user 25.9 s, sys: 11.6 s, total: 37.6 s
Wall time: 42.2 s



## Results Summary Table

| # | Method | Model | Steps | Gen Length | Block | Wall Time (s) | Time/Sample (s) | Time/Generation (s) | Throughput (gen/s) | Output Quality | Status |
|---|--------|-------|-------|------------|-------|---------------|-----------------|---------------------|-------------------|----------------|--------|
| 1 | **Official LLaDA** | LLaDA-8B-Base | **256** | 256 | 32 | 49.05 | 4.90 | 2.45 | 0.41 | ✅ **Excellent** | **WORKING** |
| 2 | **Fast-dLLM** | LLaDA-8B-Base | 256 | 256 | 32 | 7.90 | 0.79 | 0.40 | 2.53 | ⚠️ Good | Minor Issues |
| 3 | **Official LLaDA** | LLaDA-8B-Base | 64 | 256 | 32 | 12.58 | 1.26 | 0.63 | 1.59 | ❌ Failed | Zeros/incomplete |
| 4 | **Official LLaDA** | LLaDA-8B-Base | 32 | 256 | 32 | 6.09 | 0.61 | 0.30 | 3.28 | ❌ Failed | Completely garbled |
| 5 | **Autoregressive** | Qwen2.5-7B | N/A | 256 | N/A | 11.37 | 1.14 | 0.57 | 1.76 | ✅ **Excellent** | **WORKING** |

In [5]:
%%time

diffusion_model_name = "Dream-org/Dream-v0-Base-7B"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 32
temperature = 0.0
cfg_scale = 0.0
steps = 32
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2

metrics = evaluate_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples
    )

Loading checkpoint shards: 100%|██████████| 4/4 [00:03<00:00,  1.04it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [00:13<00:00,  4.57s/it]


Saved generations to results/Dream-org_Dream-v0-Base-7B_256_32_0_2_0.0_generations_testing.json
CPU times: user 16.7 s, sys: 17.1 s, total: 33.8 s
Wall time: 51.3 s


In [6]:
%%time

diffusion_model_name = "Dream-org/Dream-v0-Base-7B"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 64
temperature = 0.0
cfg_scale = 0.0
steps = 64
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2

metrics = evaluate_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples
    )

Loading checkpoint shards: 100%|██████████| 4/4 [00:00<00:00, 102.54it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [00:30<00:00, 10.06s/it]


Saved generations to results/Dream-org_Dream-v0-Base-7B_256_64_0_2_0.0_generations_testing.json
CPU times: user 32.5 s, sys: 9.74 s, total: 42.2 s
Wall time: 46.8 s


In [7]:
%%time

diffusion_model_name = "Dream-org/Dream-v0-Base-7B"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 2
gen_length = 256
diffusion_steps = 256
temperature = 0.0
cfg_scale = 0.0
steps = 256
block_length = 32
remasking = "low_confidence"
alg = "entropy"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2

metrics = evaluate_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples
    )

Loading checkpoint shards: 100%|██████████| 4/4 [00:00<00:00, 95.21it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 3/3 [01:59<00:00, 39.73s/it]


Saved generations to results/Dream-org_Dream-v0-Base-7B_256_256_0_2_0.0_generations_testing.json
CPU times: user 2min 1s, sys: 10.8 s, total: 2min 12s
Wall time: 2min 18s


In [ ]:
%%time

diffusion_model_name = "Dream-org/Dream-v0-Base-7B"
data = "gsm8k"
num_evals_to_use = 5
few_shot = 0
batch_size = 1
gen_length = 256
diffusion_steps = 256
temperature = 0.0
cfg_scale = 0.0
steps = 256
block_length = 32
remasking = "low_confidence"
alg = "confidence_threshold"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 2
use_cache=True
dual_cache=True
threshold = 0.75
factor = 1.0

metrics = evaluate_fast_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples, use_cache, dual_cache, threshold, factor
    )

Loading checkpoint shards: 100%|██████████| 4/4 [00:00<00:00, 99.64it/s]


evaluating 5 examples
Model type: Base


Evaluating (total samples: 5): 100%|██████████| 5/5 [00:33<00:00,  6.63s/it]


Saved generations to results/Dream-org_Dream-v0-Base-7B_256_256_0_2_0.0_generations_testing_fast_dllm.json
CPU times: user 35.2 s, sys: 9.64 s, total: 44.9 s
Wall time: 49.6 s


## Results Summary Table (End-to-End Times)

| # | Method | Model | Steps | Temp | Block | Cache | Alg | Wall Time (s) | Time/Sample (s) | Time/Gen (s) | Throughput (tok/s) | Output Quality | Status |
|---|--------|-------|-------|------|-------|-------|-----|---------------|-----------------|-----------------|-------------------|----------------|--------|
| 1 | **Dream Official** | Dream-7B | 256 | 0.0 | N/A | ❌ No | entropy | **138.0s** | 13.80 | 6.90 | 37.1 tok/s | ⚠️ Mixed | Incomplete |
| 2 | **Dream Official** | Dream-7B | 64 | 0.0 | N/A | ❌ No | entropy | **46.8s** | 4.68 | 2.34 | 109.4 tok/s | ⚠️ Mixed | Incomplete |
| 3 | **Dream Official** | Dream-7B | 32 | 0.0 | N/A | ❌ No | entropy | **51.3s** | 5.13 | 2.57 | 99.9 tok/s | ❌ Failed | Loading overhead |
| 4 | **Fast-dLLM** | Dream-7B | 256 | 0.0 | 32 | ✅ Dual | entropy | **49.6s** | 4.96 | 2.48 | 103.2 tok/s | ❌ Poor | Mask tokens |
| 5 | **Autoregressive** | Qwen-7B | N/A | 0.0 | N/A | ❌ No | N/A | **55.6s** | 5.56 | 2.78 | 92.1 tok/s | ✅ **Perfect** | **WORKING** |

**Note**: These are end-to-end times including model loading overhead

In [ ]:
%%time

diffusion_model_name = "Dream-org/Dream-v0-Base-7B"
data = "gsm8k"
num_evals_to_use = 256
few_shot = 4
batch_size = 1
gen_length = 256
diffusion_steps = 256
temperature = 0.0
cfg_scale = 0.0
steps = 256
block_length = 32
remasking = "low_confidence"
alg = "confidence_threshold"
alg_temp = 0.0
top_p = 0.95
top_k = None
n_samples = 1
use_cache=True
dual_cache=True
threshold = 0.75
factor = 1.0

metrics = evaluate_fast_dllm(
    diffusion_model_name, data, num_evals_to_use, few_shot, batch_size, gen_length, diffusion_steps, temperature, cfg_scale, steps, block_length, remasking, alg, alg_temp, top_p, top_k, n_samples, use_cache, dual_cache, threshold, factor
    )

Loading checkpoint shards: 100%|██████████| 4/4 [00:00<00:00, 103.34it/s]


Created 4 few-shot examples
evaluating 256 examples
Model type: Base


Evaluating (total samples: 256):  28%|██▊       | 72/256 [06:47<18:23,  6.00s/it]